Implement #8057 103 Early Hint

gregw opened the pull request:
Implementing #8057 Early hints

gregw committed:
Implement #8057 103 Early Hint

Added test harness for intermediary responses.

gregw committed:
Implement #8057 103 Early Hint

Implemented 103 early hint, but is failing for HTTP>=2

gregw committed:
Implement #8057 103 Early Hint

Fixed H2
Attempted fix H3

gregw committed:
Implement #8057 103 Early Hint

Fixed H2
Attempted fix H3

gregw commented:
@lorban can you look at why h3 is failing.

lorban commented:
@gregw I see no H3 test failure in the PR's build, only a checkstyle error. What am I overlooking?

gregw commented:
@lorban have you run org.eclipse.jetty.http.client.IntermediateResponseTest#test103EarlyHint ?

gregw committed:
Implement #8057 103 Early Hint

fixed checkstyle

lorban committed:
add missing H3 support for status 103

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban commented:
@gregw I've added the missing support in H3 for 103 status code and pushed the changes to this branch.

While your test now passes and I'm reasonably confident my changes are correct, it would still be wise to have @sbordet review this bit eventually.

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/HttpGenerator.java:
```diff
@@ -388,12 +388,18 @@ public Result generateResponse(MetaData.Response info, boolean head, ByteBuffer
                     if (status >= 100 && status < 200)
```
There is `HttpStatus.isInformational()` that we should use here.

sbordet commented on jetty-http2/http2-server/src/main/java/org/eclipse/jetty/http2/server/HttpTransportOverHTTP2.java:
```diff
@@ -98,7 +98,7 @@ public void sendHeaders(MetaData.Request request, MetaData.Response response, By
         boolean hasContent = BufferUtil.hasContent(content) && !isHeadRequest;
         int status = response.getStatus();
-        boolean interimResponse = status == HttpStatus.CONTINUE_100 || status == HttpStatus.PROCESSING_102;
+        boolean interimResponse = status >= HttpStatus.CONTINUE_100 && status < HttpStatus.OK_200 && status != HttpStatus.SWITCHING_PROTOCOLS_101;
```
Replace with `HttpStatus.isInformational() && !101`.

sbordet commented on tests/test-http-client-transport/src/test/java/org/eclipse/jetty/http/client/IntermediateResponseTest.java:
```diff
@@ -0,0 +1,298 @@
+
+    @ParameterizedTest
+    @ArgumentsSource(TransportProvider.class)
+    public void test100Continue(Transport transport) throws Exception
```
There are a lot of 100-Continue tests elsewhere, so I don't think we need to duplicate this here.
I'd remove this test.

gregw replied:
I'm not sure the other tests made sending the content actually conditional on receiving the 100, as I certainly could not find the magic needed in the HttpClient elsewhere.    So I'd be more inclined to delete the other tests and keep this one.  I'll look for any other differences

gregw replied:
OK now I find the good 100 test!

sbordet commented on tests/test-http-client-transport/src/test/java/org/eclipse/jetty/http/client/IntermediateResponseTest.java:
```diff
@@ -0,0 +1,298 @@
+                        if (exchange != null && response.getStatus() == HttpStatus.EARLY_HINT_103)
+                        {
+                            // All good, continue.
+                            System.err.println("onSuccess\n" + response.getHeaders());
```
Remove `System.err`

sbordet commented on tests/test-http-client-transport/src/test/java/org/eclipse/jetty/http/client/IntermediateResponseTest.java:
```diff
@@ -0,0 +1,298 @@
+                        }
+                        else
+                        {
+                            throw new IllegalStateException("should not have accepted");
```
This exception will only be logged, better call `response.abort(new ISE(...))`, so the test will fail.

sbordet commented on tests/test-http-client-transport/src/test/java/org/eclipse/jetty/http/client/IntermediateResponseTest.java:
```diff
@@ -0,0 +1,298 @@
+                        }
+                        else
+                        {
+                            throw new IllegalStateException("should not have accepted");
```
Use `response.abort(new ISE(...))`.

sbordet commented on jetty-http3/http3-common/src/main/java/org/eclipse/jetty/http3/internal/HTTP3StreamConnection.java:
```diff
@@ -486,7 +486,7 @@ public void onHeaders(long streamId, HeadersFrame frame)
             {
                 MetaData.Response response = (MetaData.Response)metaData;
-                if (response.getStatus() != HttpStatus.CONTINUE_100)
+                if (response.getStatus() != HttpStatus.CONTINUE_100 && response.getStatus() != HttpStatus.EARLY_HINT_103)
```
This should be `HttpStatus.isInformational()`.

gregw replied:
Is that OK for a 101 switching protocols?

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/HttpStatus.java:
```diff
@@ -103,6 +104,7 @@
         CONTINUE(CONTINUE_100, "Continue"),
         SWITCHING_PROTOCOLS(SWITCHING_PROTOCOLS_101, "Switching Protocols"),
         PROCESSING(PROCESSING_102, "Processing"),
+        EARLY_HINT(EARLY_HINT_103, "Early Hint"),
```
Let's add also `HttpHeader.Link` from RFC 8288.

gregw replied:
Added in javadoc

sbordet commented on jetty-http3/http3-client/src/main/java/org/eclipse/jetty/http3/client/internal/HTTP3StreamClient.java:
```diff
@@ -53,6 +53,8 @@ public void onResponse(HeadersFrame frame)
         if (response.getStatus() == HttpStatus.CONTINUE_100)
             valid = validateAndUpdate(EnumSet.of(FrameState.INITIAL), FrameState.CONTINUE);
+        else if (response.getStatus() == HttpStatus.EARLY_HINT_103)
+            valid = validateAndUpdate(EnumSet.of(FrameState.INITIAL, FrameState.HEADER, FrameState.CONTINUE), FrameState.CONTINUE);
```
I would rename `FrameState.CONTINUE` to `INFORMATIONAL`.

State HEADER is the state used to validate data frames (i.e. we can only receive a DATA frame if we are in HEADER state).
So I think the "from" states can only be INITIAL and CONTINUE(INFORMATIONAL).

lorban replied:
I agree that `FrameState.CONTINUE` should be renamed to `FrameState.INFORMATIONAL`.

I'm wondering if this code could be generalized or if we have no choice but to stick to a `if (status == 103)` special case.

gregw replied:
There is no special case for 102???  Let's hangout and walk through this logic.

sbordet commented on jetty-http3/http3-server/src/main/java/org/eclipse/jetty/http3/server/internal/HttpTransportOverHTTP3.java:
```diff
@@ -71,7 +71,7 @@ private void sendHeaders(MetaData.Request request, MetaData.Response response, B
         boolean hasContent = BufferUtil.hasContent(content) && !isHeadRequest;
         int status = response.getStatus();
-        boolean interimResponse = status == HttpStatus.CONTINUE_100 || status == HttpStatus.PROCESSING_102;
+        boolean interimResponse = status >= HttpStatus.CONTINUE_100 && status < HttpStatus.OK_200 && status != HttpStatus.SWITCHING_PROTOCOLS_101;
```
Replace with `HttpStatus.isInformational() && !101`.

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/Response.java:
```diff
@@ -514,6 +517,23 @@ public void sendProcessing() throws IOException
+     *
+     * @throws IOException if unable to send the 102 response
+     * @see javax.servlet.http.HttpServletResponse#sendError(int)
+     */
```
Javadoc is wrong.

sbordet commented on tests/test-http-client-transport/src/test/java/org/eclipse/jetty/http/client/IntermediateResponseTest.java:
```diff
@@ -0,0 +1,298 @@
+import static org.hamcrest.Matchers.is;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class IntermediateResponseTest extends AbstractTest<TransportScenario>
```
Rename to `InformationalResponseTest`.

sbordet commented on tests/test-http-client-transport/src/test/java/org/eclipse/jetty/http/client/IntermediateResponseTest.java:
```diff
@@ -0,0 +1,298 @@
+                    @Override
+                    public void onSuccess(Response response)
+                    {
+                        org.eclipse.jetty.client.api.Request request = response.getRequest();
```
Use `var request = ...`.

sbordet commented on tests/test-http-client-transport/src/test/java/org/eclipse/jetty/http/client/IntermediateResponseTest.java:
```diff
@@ -0,0 +1,298 @@
+                    @Override
+                    public void onSuccess(Response response)
+                    {
+                        org.eclipse.jetty.client.api.Request request = response.getRequest();
```
Use `var request = ...`.

gregw committed:
Implement #8057 103 Early Hint

updates from review

gregw commented:
@lorban @sbordet I've updated for review feedback.
However, there is still something racy either in the tests or the impl.   UNIX_DOMAIN frequently fails for me if run with other tests (the second 103 is skipped).  I have once seen H2 fail by the 103 being reported as the final response instead of the 200, but I can't reproduce.     Something is not quiet right :(

gregw commented:
Hmmm and I've just had a flake of the 102 test for UNIX_DOMAIN as well... also not reproducible :(

gregw commented:
@sbordet what does the client do if it is still processing the 102/103 response and the final (or next) response turns up?    On the server, we don't wait for anything from the client before sending the next response, but if I put in a 10ms sleep between the responses then I don't get any flakes.      Either we are writing over the responses or the client is processing the next one before the last is finished?

gregw commented:
@sbordet I think the race is in the client.  If I put the 10ms sleeps in the server I pass every time. However, if I put a 100ms sleep in the onSuccess handling for the 1xx responses, then I fail more often than not, even for HTTP.

lorban commented on jetty-server/src/main/java/org/eclipse/jetty/server/Response.java:
```diff
@@ -514,6 +517,22 @@ public void sendProcessing() throws IOException
+     * Send a 103 response as per <a href="https://datatracker.ietf.org/doc/html/rfc8297">RFC8297</a>
+     * This method is called by sendError if it is passed 103.
+     *
+     * @throws IOException if unable to send the 102 response
```
```diff
- if unable to send the 102 response
+ if unable to send the 103 response
```

lorban commented:
All tests in `IntermediateResponseTest` eventually fail when ran enough times, even if the H3 and UNIX_DOMAIN ones are more sensible.

gregw commented:
@sbordet @lorban The flaky issue is somewhere below the generation of the response in the server and the parsing of it in the client.   I added some println's and for a good run I see:
```
generateResponseLine HTTP/1.1{s=103,h=2,cl=-1}
startResponse 103 Early Hint
generateResponseLine HTTP/1.1{s=103,h=2,cl=-1}
startResponse 103 Early Hint
generateResponseLine HTTP/1.1{s=200,h=2,cl=-1}
startResponse 200 OK
```
So each generated response is parsed (startResponse) in the client.

But for a bad run I see:
```
generateResponseLine HTTP/1.1{s=103,h=2,cl=-1}
generateResponseLine HTTP/1.1{s=103,h=2,cl=-1}
startResponse 103 Early Hint
generateResponseLine HTTP/1.1{s=200,h=2,cl=-1}
startResponse 200 OK
```
So all the responses are generated, but one of them is not received.     It feels like something might be clearing a buffer when it should not?     

@lorban can you focus on the H2/H3 flakes and I'll keep looking at h1

gregw committed:
Implement #8057 103 Early Hint

updates from review

gregw commented:
oops sorry forced pushed after a rebase

gregw committed:
Implement #8057 103 Early Hint

Found (but not fixed) flake.

gregw commented on jetty-client/src/main/java/org/eclipse/jetty/client/http/HttpReceiverOverHTTP.java:
```diff
@@ -252,6 +252,7 @@ private boolean parse()
 
             if (complete)
             {
+                // TODO this can clean a pipelined response after a 1xx
                 if (LOG.isDebugEnabled())
                     LOG.debug("Discarding unexpected content after response: {}", networkBuffer);
                 networkBuffer.clear();
```
@sbordet @lorban this is the cause of the UNIX_DOMAIN flake.  A pipelined response can be discarded after a 1xx response.    I'm not sure of the fix.

gregw committed:
Implement #8057 103 Early Hint

Fixed flake. @sbordet please review this commit!

gregw commented:
@sbordet I found and maybe fixed the flake in the HttpClient (discarding content after a 1xx), but would appreciate very careful review.
@lorban I can't get any flaky fails in h2 or h3? How about you?

lorban commented:
@gregw I can make `HttpClientStreamTest.testFileUpload` on H3 fail about once every couple dozen runs (I'm aware of that failure and I already started investigating it) but I cannot manage to get any of the tests in `InformationalResponseTest` to fail for H2 or H3 anymore.

That's strange...

sbordet commented:
@gregw looking

sbordet committed:
Issue #8057 - Support Http Response 103 (Early Hints)

Improved implementation on the client-side.
Introduced HttpStatus.isInterim() for 1XX codes that are not 101.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>
